pkg/resourcemanager/webhook/highavailabilityconfig/handler.go

```diff
@@ -420,24 +420,24 @@ func (h *Handler) mutateNodeAffinity(
 			podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution = &corev1.NodeSelector{}
 		}
 
-		// Filter existing terms with the same expression key to prevent that we are trying to add an expression with
-		// the same key multiple times.
-		var filteredNodeSelectorTerms []corev1.NodeSelectorTerm
-		for _, term := range podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		if len(podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) == 0 {
+			podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms = []corev1.NodeSelectorTerm{{}}
+		}
+
+		nodeSelectorTerms := podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
+		for i, term := range nodeSelectorTerms {
+			filteredExpressions := make([]corev1.NodeSelectorRequirement, 0, len(term.MatchExpressions))
+			// Remove existing expressions for `topology.kubernetes.io/zone` to
+			// - avoid duplicates for the same key
+			// - remove expressions that prevent zone pinning
 			for _, expr := range term.MatchExpressions {
 				if expr.Key != corev1.LabelTopologyZone {
-					filteredNodeSelectorTerms = append(filteredNodeSelectorTerms, term)
+					filteredExpressions = append(filteredExpressions, expr)
 				}
 			}
-		}
-		podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms = filteredNodeSelectorTerms
-
-		if len(podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) == 0 {
-			podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms = []corev1.NodeSelectorTerm{{}}
-		}
 
-		for i := range podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
-			podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[i].MatchExpressions = append(podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[i].MatchExpressions, *nodeSelectorRequirement)
+			// Add remaining expressions with intended zone expression.
+			nodeSelectorTerms[i].MatchExpressions = append(filteredExpressions, *nodeSelectorRequirement)
 		}
```

Comment:
After updating the nodeSelectorTerms, we are not assigning it back to `podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms`

```suggestion
		}
		podTemplateSpec.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms = nodeSelectorTerms
```

Reply:
Don't they share the same underlying data/slice? Otherwise the integration test would fail.

Reply:
Yes, but this is unintuitive.
Also [this](https://go.dev/play/p/DnsK3wy45LU) kind of case doesn't go well. IMO its better to do the assignment at the later part also.
But anything is fine.